table checkpoint-ts can be larger than redo-resolved-ts

### What did you do?

Here is a piece of TiCDC log:
```
egv-transactions-ch2-ticdc-2 ticdc [2023/09/15 20:04:27.713 +05:30] [INFO] [manager.go:850] ["Start table sink"] [namespace=default] [changefeed=egv-transactions-1-hyd] [tableID=74] [startTs=444278613643624570]
egv-transactions-ch2-ticdc-2 ticdc [2023/09/15 20:04:27.713 +05:30] [INFO] [table_sink_wrapper.go:145] ["Sink is started"] [namespace=default] [changefeed=egv-transactions-1-hyd] [tableID=74] [startTs=444278613643624570] [replicateTs=444278628074651689]
egv-transactions-ch2-ticdc-2 ticdc [2023/09/15 20:04:27.713 +05:30] [INFO] [processor.go:325] ["Add Table finished"] [captureID=18f5d46d-f065-4214-b118-b63dd3dc4fb9] [namespace=default] [changefeed=egv-transactions-1-hyd] [tableID=74] [tableResolvedTs=444278613643624570] [tableCheckpointTs=444278613643624570] [globalCheckpointTs=444278613643624570] [state=Replicating] [isPrepare=false]
egv-transactions-ch2-ticdc-2 ticdc [2023/09/15 20:04:27.813 +05:30] [ERROR] [manager.go:1019] ["sinkManager: resolved ts should not less than checkpoint ts"] [namespace=default] [changefeed=egv-transactions-1-hyd] [tableID=74] [resolvedTs=444278613643624570] [checkpointTs="{\"Mode\":0,\"Ts\":444278627681435676,\"BatchID\":18446744073709551615}"]
```

### What did you expect to see?

_No response_

### What did you see instead?

For a given table, checkpoint-ts is larger than redo-resolved-ts.
This check happens at `cdc/processor/sinkmanager/manager.go:1019`.
TiCDC version is `v6.5.4`.

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
(paste TiDB cluster version here)
```

Upstream TiKV version (execute `tikv-server --version`):

```console
(paste TiKV version here)
```

TiCDC version (execute `cdc version`):

```console
(paste TiCDC version here)
```

It's reproduced. Seems lots of tables (more than 200) and more than one captures are also required.

```
[2023/09/20 13:33:44.178 +08:00] [ERROR] [processor.go:646] ["run processor failed"] [capture=668dd527-ebd6-4f63-b2d9-313966635325] [namespace=default] [changefeed=x] [erro
r="[CDC:ErrInvalidCheckpointTs]checkpointTs(444383368798535681) should not larger than resolvedTs(444383368052736000)"] [errorVerbose="[CDC:ErrInvalidCheckpointTs]checkpoin
tTs(444383368798535681) should not larger than resolvedTs(444383368052736000)\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20220729040631-518f
63d66278/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20220729040631-518f63d66278/normalize.go:164\ngithub.co
m/pingcap/tiflow/cdc/scheduler/internal/v3/agent.(*agent).handleMessageHeartbeat\n\tgithub.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent/agent.go:266\ngithub.com/pingc
ap/tiflow/cdc/scheduler/internal/v3/agent.(*agent).handleMessage\n\tgithub.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent/agent.go:234\ngithub.com/pingcap/tiflow/cdc/sc
heduler/internal/v3/agent.(*agent).Tick\n\tgithub.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent/agent.go:186\ngithub.com/pingcap/tiflow/cdc/processor.(*processor).tick
\n\tgithub.com/pingcap/tiflow/cdc/processor/processor.go:698\ngithub.com/pingcap/tiflow/cdc/processor.(*processor).Tick\n\tgithub.com/pingcap/tiflow/cdc/processor/processor
.go:592\ngithub.com/pingcap/tiflow/cdc/processor.(*managerImpl).Tick\n\tgithub.com/pingcap/tiflow/cdc/processor/manager.go:133\ngithub.com/pingcap/tiflow/pkg/orchestrator.(
*EtcdWorker).Run\n\tgithub.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:290\ngithub.com/pingcap/tiflow/cdc/capture.(*captureImpl).runEtcdWorker\n\tgithub.com/pingcap/
tiflow/cdc/capture/capture.go:548\ngithub.com/pingcap/tiflow/cdc/capture.(*captureImpl).run.func4\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:380\ngolang.org/x/sync
/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.1.0/errgroup/errgroup.go:75\nruntime.goexit\n\truntime/asm_amd64.s:1650"]
[2023/09/20 13:33:53.891 +08:00] [ERROR] [feed_state_manager.go:452] ["processor reports an error"] [namespace=default] [changefeed=x] [captureID=668dd527-ebd6-4f63-b2d9-31
3966635325] [error="{\"time\":\"2023-09-20T13:33:53.889723519+08:00\",\"addr\":\"10.2.7.2:8300\",\"code\":\"CDC:ErrInvalidCheckpointTs\",\"message\":\"[CDC:ErrInvalidCheckp
ointTs]checkpointTs(444383368798535681) should not larger than resolvedTs(444383368052736000)\"}"]
```

--- a/cdc/processor/sinkmanager/manager.go
+++ b/cdc/processor/sinkmanager/manager.go
@@ -771,7 +771,8 @@ func (m *SinkManager) UpdateReceivedSorterResolvedTs(span tablepb.Span, ts model
 func (m *SinkManager) UpdateBarrierTs(globalBarrierTs model.Ts, tableBarrier map[model.TableID]model.Ts) {
 	m.tableSinks.Range(func(span tablepb.Span, value interface{}) bool {
 		barrierTs := globalBarrierTs
-		if tableBarrierTs, ok := tableBarrier[span.TableID]; ok && tableBarrierTs > barrierTs {
+		if tableBarrierTs, ok := tableBarrier[span.TableID]; ok && tableBarrierTs < barrierTs {
+			// TODO: is it possible that table barrier is less than global barrier?
 			barrierTs = tableBarrierTs
 		}
 		value.(*tableSinkWrapper).updateBarrierTs(barrierTs)

--- a/cdc/processor/sinkmanager/manager.go
+++ b/cdc/processor/sinkmanager/manager.go
@@ -772,8 +772,10 @@ func (m *SinkManager) UpdateBarrierTs(globalBarrierTs model.Ts, tableBarrier map
 	m.tableSinks.Range(func(span tablepb.Span, value interface{}) bool {
 		barrierTs := globalBarrierTs
 		if tableBarrierTs, ok := tableBarrier[span.TableID]; ok && tableBarrierTs < barrierTs {
-			// TODO: is it possible that table barrier is less than global barrier?
-			barrierTs = tableBarrierTs
+			log.Panic("tableBarrierTs should be always greater than globalBarrierTs",
+				zap.String("namespace", m.changefeedID.Namespace),
+				zap.String("changefeed", m.changefeedID.ID),
+				zap.Stringer("span", &span))
 		}
 		value.(*tableSinkWrapper).updateBarrierTs(barrierTs)
 		return true

--- a/cdc/processor/sinkmanager/manager.go
+++ b/cdc/processor/sinkmanager/manager.go
@@ -768,30 +768,13 @@ func (m *SinkManager) UpdateReceivedSorterResolvedTs(span tablepb.Span, ts model
 }
 
 // UpdateBarrierTs update all tableSink's barrierTs in the SinkManager
-func (m *SinkManager) UpdateBarrierTs(
-	globalBarrierTs model.Ts,
-	tableBarrier map[model.TableID]model.Ts,
-) {
+func (m *SinkManager) UpdateBarrierTs(globalBarrierTs model.Ts, tableBarrier map[model.TableID]model.Ts) {
 	m.tableSinks.Range(func(span tablepb.Span, value interface{}) bool {
-		tableSink := value.(*tableSinkWrapper)
-		lastBarrierTs := tableSink.barrierTs.Load()
-		// It is safe to do not use compare and swap here.
-		// Only the processor will update the barrier ts.
-		// Other goroutines will only read the barrier ts.
-		// So it is safe to do not use compare and swap here, just Load and Store.
-		if tableBarrierTs, ok := tableBarrier[tableSink.span.TableID]; ok {
-			barrierTs := tableBarrierTs
-			if barrierTs > globalBarrierTs {
-				barrierTs = globalBarrierTs
-			}
-			if barrierTs > lastBarrierTs {
-				tableSink.barrierTs.Store(barrierTs)
-			}
-		} else {
-			if globalBarrierTs > lastBarrierTs {
-				tableSink.barrierTs.Store(globalBarrierTs)
-			}
+		barrierTs := globalBarrierTs
+		if tableBarrierTs, ok := tableBarrier[span.TableID]; ok && tableBarrierTs > barrierTs {
+			barrierTs = tableBarrierTs
 		}
+		value.(*tableSinkWrapper).updateBarrierTs(barrierTs)
 		return true
 	})
 }
@@ -1008,14 +991,14 @@ func (m *SinkManager) GetTableStats(span tablepb.Span) TableStats {
 		resolvedTs = tableSink.getReceivedSorterResolvedTs()
 	}
 
-	if resolvedTs < checkpointTs.ResolvedMark() {
-		log.Error("sinkManager: resolved ts should not less than checkpoint ts",
+	sinkUpperBound := tableSink.getUpperBoundTs()
+	if sinkUpperBound < checkpointTs.ResolvedMark() {
+		log.Panic("sinkManager: sink upperbound should not less than checkpoint ts",
 			zap.String("namespace", m.changefeedID.Namespace),
 			zap.String("changefeed", m.changefeedID.ID),
 			zap.Stringer("span", &span),
-			zap.Uint64("resolvedTs", resolvedTs),
-			zap.Any("checkpointTs", checkpointTs),
-			zap.Uint64("barrierTs", tableSink.barrierTs.Load()))
+			zap.Uint64("upperbound", sinkUpperBound),
+			zap.Any("checkpointTs", checkpointTs))
 	}
 	return TableStats{
 		CheckpointTs: checkpointTs.ResolvedMark(),

--- a/cdc/processor/sinkmanager/manager.go
+++ b/cdc/processor/sinkmanager/manager.go
@@ -771,11 +771,8 @@ func (m *SinkManager) UpdateReceivedSorterResolvedTs(span tablepb.Span, ts model
 func (m *SinkManager) UpdateBarrierTs(globalBarrierTs model.Ts, tableBarrier map[model.TableID]model.Ts) {
 	m.tableSinks.Range(func(span tablepb.Span, value interface{}) bool {
 		barrierTs := globalBarrierTs
-		if tableBarrierTs, ok := tableBarrier[span.TableID]; ok && tableBarrierTs < barrierTs {
-			log.Panic("tableBarrierTs should be always greater than globalBarrierTs",
-				zap.String("namespace", m.changefeedID.Namespace),
-				zap.String("changefeed", m.changefeedID.ID),
-				zap.Stringer("span", &span))
+		if tableBarrierTs, ok := tableBarrier[span.TableID]; ok && tableBarrierTs < globalBarrierTs {
+			barrierTs = tableBarrierTs
 		}
 		value.(*tableSinkWrapper).updateBarrierTs(barrierTs)
 		return true

--- a/cdc/processor/sinkmanager/table_sink_wrapper.go
+++ b/cdc/processor/sinkmanager/table_sink_wrapper.go
@@ -179,6 +179,15 @@ func (t *tableSinkWrapper) appendRowChangedEvents(events ...*model.RowChangedEve
 	return nil
 }
 
+func (t *tableSinkWrapper) updateBarrierTs(ts model.Ts) {
+	for {
+		old := t.barrierTs.Load()
+		if ts <= old || t.barrierTs.CompareAndSwap(old, ts) {
+			break
+		}
+	}
+}
+
 func (t *tableSinkWrapper) updateReceivedSorterResolvedTs(ts model.Ts) {
 	for {
 		old := t.receivedSorterResolvedTs.Load()

--- a/cdc/scheduler/internal/v3/agent/agent.go
+++ b/cdc/scheduler/internal/v3/agent/agent.go
@@ -270,6 +270,12 @@ func (a *agent) handleMessageHeartbeat(request *schedulepb.Heartbeat) (*schedule
 
 	allTables.Ascend(func(span tablepb.Span, table *tableSpan) bool {
 		status := table.getTableSpanStatus(request.CollectStats)
+		if status.Checkpoint.CheckpointTs > status.Checkpoint.ResolvedTs {
+			log.Warn("schedulerv3: CheckpointTs is greater than ResolvedTs",
+				zap.String("namespace", a.ChangeFeedID.Namespace),
+				zap.String("changefeed", a.ChangeFeedID.ID),
+				zap.String("span", span.String()))
+		}
 		if table.task != nil && table.task.IsRemove {
 			status.State = tablepb.TableStateStopping
 		}

--- a/cdc/scheduler/internal/v3/agent/agent.go
+++ b/cdc/scheduler/internal/v3/agent/agent.go
@@ -275,25 +275,28 @@ func (a *agent) handleMessage(msg []*schedulepb.Message) (
 func (a *agent) handleMessageHeartbeat(request *schedulepb.Heartbeat) (
 	*schedulepb.Message, *schedulepb.Barrier, error,
 ) {
+	// NOTE: for a given table, we don't check ResolvedTs should never less than CheckpointTs
+	// because if redo is enabled, the table's ResolvedTs can be less than the global Barrier.
+	// For example, consider such steps:
+	// 1. currently CheckpointTs is 90, and RedoMeta.ResolvedTs is 100. However,
+	//    there could be a table whose RedoResolvedTs has been advanced to 110.
+	// 2. then balance the table to another processor. StartTs and CheckpointTs of the table
+	//    will become 90.
+	// 3. then RedoResolvedTs can be advanced to 110, which will also advance global BarrierTs
+	//    to 110.
+	// 4. after the just re-scheduled table finds global BarrierTs has been advanced, it can
+	//    advance its CheckpointTs to 110, although its local RedoResolvedTs is still 90.
 	allTables := a.tableM.getAllTableSpans()
 	result := make([]tablepb.TableStatus, 0, allTables.Len())
 
-	isValidCheckpointTs := true
 	allTables.Ascend(func(span tablepb.Span, table *tableSpan) bool {
 		status := table.getTableSpanStatus(request.CollectStats)
-		isValidCheckpointTs = status.Checkpoint.CheckpointTs <= status.Checkpoint.ResolvedTs
 		if table.task != nil && table.task.IsRemove {
 			status.State = tablepb.TableStateStopping
 		}
 		result = append(result, status)
-		return isValidCheckpointTs
+		return true
 	})
-	if !isValidCheckpointTs {
-		status := result[len(result)-1]
-		checkpointTs := status.Checkpoint.CheckpointTs
-		resolvedTs := status.Checkpoint.ResolvedTs
-		return nil, nil, errors.ErrInvalidCheckpointTs.GenWithStackByArgs(checkpointTs, resolvedTs)
-	}
 	for _, span := range request.GetSpans() {
 		if _, ok := allTables.Get(span); !ok {
 			status := a.tableM.getTableSpanStatus(span, request.CollectStats)

--- a/cdc/scheduler/internal/v3/agent/agent.go
+++ b/cdc/scheduler/internal/v3/agent/agent.go
@@ -270,17 +270,6 @@ func (a *agent) handleMessage(msg []*schedulepb.Message) (
 }
 
 func (a *agent) handleMessageHeartbeat(request *schedulepb.Heartbeat) (*schedulepb.Message, *schedulepb.Barrier) {
-	// NOTE: for a given table, we don't check ResolvedTs should never less than CheckpointTs
-	// because if redo is enabled, the table's ResolvedTs can be less than the global Barrier.
-	// For example, consider such steps:
-	// 1. currently CheckpointTs is 90, and RedoMeta.ResolvedTs is 100. However,
-	//    there could be a table whose RedoResolvedTs has been advanced to 110.
-	// 2. then balance the table to another processor. StartTs and CheckpointTs of the table
-	//    will become 90.
-	// 3. then RedoResolvedTs can be advanced to 110, which will also advance global BarrierTs
-	//    to 110.
-	// 4. after the just re-scheduled table finds global BarrierTs has been advanced, it can
-	//    advance its CheckpointTs to 110, although its local RedoResolvedTs is still 90.
 	allTables := a.tableM.getAllTableSpans()
 	result := make([]tablepb.TableStatus, 0, allTables.Len())
 

--- a/cdc/scheduler/internal/v3/agent/agent.go
+++ b/cdc/scheduler/internal/v3/agent/agent.go
@@ -206,10 +206,7 @@ func (a *agent) Tick(ctx context.Context) (*schedulepb.Barrier, error) {
 		return nil, errors.Trace(err)
 	}
 
-	outboundMessages, barrier, err := a.handleMessage(inboundMessages)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
+	outboundMessages, barrier := a.handleMessage(inboundMessages)
 
 	responses, err := a.tableM.poll(ctx)
 	if err != nil {
@@ -237,9 +234,7 @@ func (a *agent) handleLivenessUpdate(liveness model.Liveness) {
 	}
 }
 
-func (a *agent) handleMessage(msg []*schedulepb.Message) (
-	result []*schedulepb.Message, barrier *schedulepb.Barrier, err error,
-) {
+func (a *agent) handleMessage(msg []*schedulepb.Message) (result []*schedulepb.Message, barrier *schedulepb.Barrier) {
 	for _, message := range msg {
 		ownerCaptureID := message.GetFrom()
 		header := message.GetHeader()

--- a/cdc/scheduler/internal/v3/agent/agent.go
+++ b/cdc/scheduler/internal/v3/agent/agent.go
@@ -254,10 +254,7 @@ func (a *agent) handleMessage(msg []*schedulepb.Message) (
 		switch message.GetMsgType() {
 		case schedulepb.MsgHeartbeat:
 			var reMsg *schedulepb.Message
-			reMsg, barrier, err = a.handleMessageHeartbeat(message.GetHeartbeat())
-			if err != nil {
-				return
-			}
+			reMsg, barrier = a.handleMessageHeartbeat(message.GetHeartbeat())
 			result = append(result, reMsg)
 		case schedulepb.MsgDispatchTableRequest:
 			a.handleMessageDispatchTableRequest(message.DispatchTableRequest, processorEpoch)
@@ -272,9 +269,7 @@ func (a *agent) handleMessage(msg []*schedulepb.Message) (
 	return
 }
 
-func (a *agent) handleMessageHeartbeat(request *schedulepb.Heartbeat) (
-	*schedulepb.Message, *schedulepb.Barrier, error,
-) {
+func (a *agent) handleMessageHeartbeat(request *schedulepb.Heartbeat) (*schedulepb.Message, *schedulepb.Barrier) {
 	// NOTE: for a given table, we don't check ResolvedTs should never less than CheckpointTs
 	// because if redo is enabled, the table's ResolvedTs can be less than the global Barrier.
 	// For example, consider such steps:
@@ -323,7 +318,7 @@ func (a *agent) handleMessageHeartbeat(request *schedulepb.Heartbeat) (
 		zap.String("changefeed", a.ChangeFeedID.ID),
 		zap.Any("message", message))
 
-	return message, request.GetBarrier(), nil
+	return message, request.GetBarrier()
 }
 
 type dispatchTableTaskStatus int32

--- a/cdc/scheduler/internal/v3/agent/agent_test.go
+++ b/cdc/scheduler/internal/v3/agent/agent_test.go
@@ -356,7 +356,7 @@ func TestAgentHandleMessageHeartbeat(t *testing.T) {
 		},
 	}
 
-	response, _, _ := a.handleMessage([]*schedulepb.Message{heartbeat})
+	response, _ := a.handleMessage([]*schedulepb.Message{heartbeat})
 	require.Len(t, response, 1)
 	require.Equal(t, model.LivenessCaptureAlive, response[0].GetHeartbeatResponse().Liveness)
 
@@ -376,7 +376,7 @@ func TestAgentHandleMessageHeartbeat(t *testing.T) {
 	}
 
 	a.tableM.tables.GetV(spanz.TableIDToComparableSpan(1)).task = &dispatchTableTask{IsRemove: true}
-	response, _, _ = a.handleMessage([]*schedulepb.Message{heartbeat})
+	response, _ = a.handleMessage([]*schedulepb.Message{heartbeat})
 	result = response[0].GetHeartbeatResponse().Tables
 	sort.Slice(result, func(i, j int) bool {
 		return result[i].Span.TableID < result[j].Span.TableID
@@ -384,13 +384,13 @@ func TestAgentHandleMessageHeartbeat(t *testing.T) {
 	require.Equal(t, tablepb.TableStateStopping, result[1].State)
 
 	a.handleLivenessUpdate(model.LivenessCaptureStopping)
-	response, _, _ = a.handleMessage([]*schedulepb.Message{heartbeat})
+	response, _ = a.handleMessage([]*schedulepb.Message{heartbeat})
 	require.Len(t, response, 1)
 	require.Equal(t, model.LivenessCaptureStopping, response[0].GetHeartbeatResponse().Liveness)
 
 	a.handleLivenessUpdate(model.LivenessCaptureAlive)
 	heartbeat.Heartbeat.IsStopping = true
-	response, _, _ = a.handleMessage([]*schedulepb.Message{heartbeat})
+	response, _ = a.handleMessage([]*schedulepb.Message{heartbeat})
 	require.Equal(t, model.LivenessCaptureStopping, response[0].GetHeartbeatResponse().Liveness)
 	require.Equal(t, model.LivenessCaptureStopping, a.liveness.Load())
 }
@@ -577,7 +577,7 @@ func TestAgentHandleMessage(t *testing.T) {
 	}
 
 	// handle the first heartbeat, from the known owner.
-	response, _, _ := a.handleMessage([]*schedulepb.Message{heartbeat})
+	response, _ := a.handleMessage([]*schedulepb.Message{heartbeat})
 	require.Len(t, response, 1)
 
 	addTableRequest := &schedulepb.Message{
@@ -600,17 +600,17 @@ func TestAgentHandleMessage(t *testing.T) {
 		},
 	}
 	// wrong epoch, ignored
-	responses, _, _ := a.handleMessage([]*schedulepb.Message{addTableRequest})
+	responses, _ := a.handleMessage([]*schedulepb.Message{addTableRequest})
 	require.False(t, tableM.tables.Has(spanz.TableIDToComparableSpan(1)))
 	require.Len(t, responses, 0)
 
 	// correct epoch, processing.
 	addTableRequest.Header.ProcessorEpoch = a.Epoch
-	_, _, _ = a.handleMessage([]*schedulepb.Message{addTableRequest})
+	_, _ = a.handleMessage([]*schedulepb.Message{addTableRequest})
 	require.True(t, a.tableM.tables.Has(spanz.TableIDToComparableSpan(1)))
 
 	heartbeat.Header.OwnerRevision.Revision = 2
-	response, _, _ = a.handleMessage([]*schedulepb.Message{heartbeat})
+	response, _ = a.handleMessage([]*schedulepb.Message{heartbeat})
 	require.Len(t, response, 1)
 
 	// this should never happen in real world
@@ -624,12 +624,12 @@ func TestAgentHandleMessage(t *testing.T) {
 		From:    a.ownerInfo.ID,
 	}
 
-	response, _, _ = a.handleMessage([]*schedulepb.Message{unknownMessage})
+	response, _ = a.handleMessage([]*schedulepb.Message{unknownMessage})
 	require.Len(t, response, 0)
 
 	// staled message
 	heartbeat.Header.OwnerRevision.Revision = 1
-	response, _, _ = a.handleMessage([]*schedulepb.Message{heartbeat})
+	response, _ = a.handleMessage([]*schedulepb.Message{heartbeat})
 	require.Len(t, response, 0)
 }
 

--- a/cdc/scheduler/internal/v3/replication/replication_manager.go
+++ b/cdc/scheduler/internal/v3/replication/replication_manager.go
@@ -512,24 +512,25 @@ func (r *Manager) AdvanceCheckpoint(
 	barrier *schedulepb.BarrierWithMinTs,
 	redoMetaManager redo.MetaManager,
 ) (newCheckpointTs, newResolvedTs model.Ts) {
-	limitBarrierWithRedo := func(newCheckpointTs, newResolvedTs *uint64) {
+	limitBarrierWithRedo := func(newCheckpointTs, newResolvedTs uint64) (uint64, uint64) {
 		flushedMeta := redoMetaManager.GetFlushedMeta()
 		log.Debug("owner gets flushed redo meta",
 			zap.String("namespace", r.changefeedID.Namespace),
 			zap.String("changefeed", r.changefeedID.ID),
 			zap.Uint64("flushedCheckpointTs", flushedMeta.CheckpointTs),
 			zap.Uint64("flushedResolvedTs", flushedMeta.ResolvedTs))
-		if flushedMeta.ResolvedTs != 0 && flushedMeta.ResolvedTs < *newResolvedTs {
-			*newResolvedTs = flushedMeta.ResolvedTs
+		if flushedMeta.ResolvedTs != 0 && flushedMeta.ResolvedTs < newResolvedTs {
+			newResolvedTs = flushedMeta.ResolvedTs
 		}
 
-		if *newCheckpointTs > *newResolvedTs {
-			*newCheckpointTs = *newResolvedTs
+		if newCheckpointTs > newResolvedTs {
+			newCheckpointTs = newResolvedTs
 		}
 
-		if barrier.GlobalBarrierTs > *newResolvedTs {
-			barrier.GlobalBarrierTs = *newResolvedTs
+		if barrier.GlobalBarrierTs > newResolvedTs {
+			barrier.GlobalBarrierTs = newResolvedTs
 		}
+		return newCheckpointTs, newResolvedTs
 	}
 
 	newCheckpointTs, newResolvedTs = math.MaxUint64, math.MaxUint64
@@ -596,7 +597,7 @@ func (r *Manager) AdvanceCheckpoint(
 		if redoMetaManager.Enabled() {
 			// If redo is enabled, GlobalBarrierTs should be limited by redo flushed meta.
 			newResolvedTs = barrier.RedoBarrierTs
-			limitBarrierWithRedo(&newCheckpointTs, &newResolvedTs)
+			limitBarrierWithRedo(newCheckpointTs, newResolvedTs)
 		}
 		return checkpointCannotProceed, checkpointCannotProceed
 	}
@@ -649,7 +650,7 @@ func (r *Manager) AdvanceCheckpoint(
 			zap.String("changefeed", r.changefeedID.ID),
 			zap.Uint64("newCheckpointTs", newCheckpointTs),
 			zap.Uint64("newResolvedTs", newResolvedTs))
-		limitBarrierWithRedo(&newCheckpointTs, &newResolvedTs)
+		return limitBarrierWithRedo(newCheckpointTs, newResolvedTs)
 	}
 
 	return newCheckpointTs, newResolvedTs

--- a/cdc/scheduler/internal/v3/replication/replication_manager.go
+++ b/cdc/scheduler/internal/v3/replication/replication_manager.go
@@ -512,6 +512,26 @@ func (r *Manager) AdvanceCheckpoint(
 	barrier *schedulepb.BarrierWithMinTs,
 	redoMetaManager redo.MetaManager,
 ) (newCheckpointTs, newResolvedTs model.Ts) {
+	limitBarrierWithRedo := func(newCheckpointTs, newResolvedTs *uint64) {
+		flushedMeta := redoMetaManager.GetFlushedMeta()
+		log.Debug("owner gets flushed redo meta",
+			zap.String("namespace", r.changefeedID.Namespace),
+			zap.String("changefeed", r.changefeedID.ID),
+			zap.Uint64("flushedCheckpointTs", flushedMeta.CheckpointTs),
+			zap.Uint64("flushedResolvedTs", flushedMeta.ResolvedTs))
+		if flushedMeta.ResolvedTs != 0 && flushedMeta.ResolvedTs < *newResolvedTs {
+			*newResolvedTs = flushedMeta.ResolvedTs
+		}
+
+		if *newCheckpointTs > *newResolvedTs {
+			*newCheckpointTs = *newResolvedTs
+		}
+
+		if barrier.GlobalBarrierTs > *newResolvedTs {
+			barrier.GlobalBarrierTs = *newResolvedTs
+		}
+	}
+
 	newCheckpointTs, newResolvedTs = math.MaxUint64, math.MaxUint64
 	slowestRange := tablepb.Span{}
 	cannotProceed := false
@@ -575,10 +595,7 @@ func (r *Manager) AdvanceCheckpoint(
 	if cannotProceed {
 		if redoMetaManager.Enabled() {
 			// If redo is enabled, GlobalBarrierTs should be limited by redo flushed meta.
-			flushedMeta := redoMetaManager.GetFlushedMeta()
-			if barrier.GlobalBarrierTs > flushedMeta.ResolvedTs {
-				barrier.GlobalBarrierTs = flushedMeta.ResolvedTs
-			}
+			limitBarrierWithRedo(&newCheckpointTs, &newResolvedTs)
 		}
 		return checkpointCannotProceed, checkpointCannotProceed
 	}
@@ -626,26 +643,12 @@ func (r *Manager) AdvanceCheckpoint(
 			newResolvedTs = barrier.RedoBarrierTs
 		}
 		redoMetaManager.UpdateMeta(newCheckpointTs, newResolvedTs)
-		flushedMeta := redoMetaManager.GetFlushedMeta()
-		flushedCheckpointTs, flushedResolvedTs := flushedMeta.CheckpointTs, flushedMeta.ResolvedTs
-		log.Debug("owner gets flushed meta",
-			zap.Uint64("flushedResolvedTs", flushedResolvedTs),
-			zap.Uint64("flushedCheckpointTs", flushedCheckpointTs),
-			zap.Uint64("newResolvedTs", newResolvedTs),
-			zap.Uint64("newCheckpointTs", newCheckpointTs),
+		log.Debug("owner updates redo meta",
 			zap.String("namespace", r.changefeedID.Namespace),
-			zap.String("changefeed", r.changefeedID.ID))
-		if flushedResolvedTs != 0 && flushedResolvedTs < newResolvedTs {
-			newResolvedTs = flushedResolvedTs
-		}
-
-		if newCheckpointTs > newResolvedTs {
-			newCheckpointTs = newResolvedTs
-		}
-
-		if barrier.GlobalBarrierTs > newResolvedTs {
-			barrier.GlobalBarrierTs = newResolvedTs
-		}
+			zap.String("changefeed", r.changefeedID.ID),
+			zap.Uint64("newCheckpointTs", newCheckpointTs),
+			zap.Uint64("newResolvedTs", newResolvedTs))
+		limitBarrierWithRedo(&newCheckpointTs, &newResolvedTs)
 	}
 
 	return newCheckpointTs, newResolvedTs

--- a/cdc/scheduler/internal/v3/replication/replication_manager.go
+++ b/cdc/scheduler/internal/v3/replication/replication_manager.go
@@ -512,8 +512,10 @@ func (r *Manager) AdvanceCheckpoint(
 	barrier *schedulepb.BarrierWithMinTs,
 	redoMetaManager redo.MetaManager,
 ) (newCheckpointTs, newResolvedTs model.Ts) {
+	var redoFlushedResolvedTs model.Ts
 	limitBarrierWithRedo := func(newCheckpointTs, newResolvedTs uint64) (uint64, uint64) {
 		flushedMeta := redoMetaManager.GetFlushedMeta()
+		redoFlushedResolvedTs = flushedMeta.ResolvedTs
 		log.Debug("owner gets flushed redo meta",
 			zap.String("namespace", r.changefeedID.Namespace),
 			zap.String("changefeed", r.changefeedID.ID),
@@ -532,6 +534,15 @@ func (r *Manager) AdvanceCheckpoint(
 		}
 		return newCheckpointTs, newResolvedTs
 	}
+	defer func() {
+		if redoFlushedResolvedTs != 0 && barrier.GlobalBarrierTs > redoFlushedResolvedTs {
+			log.Panic("barrierTs should never greater than redo flushed",
+				zap.String("namespace", r.changefeedID.Namespace),
+				zap.String("changefeed", r.changefeedID.ID),
+				zap.Uint64("barrierTs", barrier.GlobalBarrierTs),
+				zap.Uint64("redoFlushedResolvedTs", redoFlushedResolvedTs))
+		}
+	}()
 
 	newCheckpointTs, newResolvedTs = math.MaxUint64, math.MaxUint64
 	slowestRange := tablepb.Span{}

--- a/cdc/scheduler/internal/v3/replication/replication_manager.go
+++ b/cdc/scheduler/internal/v3/replication/replication_manager.go
@@ -521,7 +521,7 @@ func (r *Manager) AdvanceCheckpoint(
 			zap.String("changefeed", r.changefeedID.ID),
 			zap.Uint64("flushedCheckpointTs", flushedMeta.CheckpointTs),
 			zap.Uint64("flushedResolvedTs", flushedMeta.ResolvedTs))
-		if flushedMeta.ResolvedTs != 0 && flushedMeta.ResolvedTs < newResolvedTs {
+		if flushedMeta.ResolvedTs < newResolvedTs {
 			newResolvedTs = flushedMeta.ResolvedTs
 		}
 

--- a/cdc/scheduler/internal/v3/replication/replication_manager.go
+++ b/cdc/scheduler/internal/v3/replication/replication_manager.go
@@ -573,6 +573,13 @@ func (r *Manager) AdvanceCheckpoint(
 		return true
 	})
 	if cannotProceed {
+		if redoMetaManager.Enabled() {
+			// If redo is enabled, GlobalBarrierTs should be limited by redo flushed meta.
+			flushedMeta := redoMetaManager.GetFlushedMeta()
+			if barrier.GlobalBarrierTs > flushedMeta.ResolvedTs {
+				barrier.GlobalBarrierTs = flushedMeta.ResolvedTs
+			}
+		}
 		return checkpointCannotProceed, checkpointCannotProceed
 	}
 	if slowestRange.TableID != 0 {

--- a/cdc/scheduler/internal/v3/replication/replication_manager.go
+++ b/cdc/scheduler/internal/v3/replication/replication_manager.go
@@ -595,6 +595,7 @@ func (r *Manager) AdvanceCheckpoint(
 	if cannotProceed {
 		if redoMetaManager.Enabled() {
 			// If redo is enabled, GlobalBarrierTs should be limited by redo flushed meta.
+			newResolvedTs = barrier.RedoBarrierTs
 			limitBarrierWithRedo(&newCheckpointTs, &newResolvedTs)
 		}
 		return checkpointCannotProceed, checkpointCannotProceed

--- a/cdc/scheduler/internal/v3/replication/replication_manager_test.go
+++ b/cdc/scheduler/internal/v3/replication/replication_manager_test.go
@@ -652,11 +652,22 @@ func TestReplicationManagerAdvanceCheckpoint(t *testing.T) {
 	require.Equal(t, model.Ts(10), checkpoint)
 	require.Equal(t, model.Ts(20), resolved)
 
-	// some table not exist yet.
+	// some table not exist yet with redo is enabled.
 	currentTables.UpdateTables([]model.TableID{1, 2, 3})
-	checkpoint, resolved = r.AdvanceCheckpoint(currentTables, time.Now(), schedulepb.NewBarrierWithMinTs(30), redoMetaManager)
+	barrier := schedulepb.NewBarrierWithMinTs(30)
+	redoMetaManager.enable = true
+	redoMetaManager.resolvedTs = 25
+	checkpoint, resolved = r.AdvanceCheckpoint(currentTables, time.Now(), barrier, redoMetaManager)
 	require.Equal(t, checkpointCannotProceed, checkpoint)
 	require.Equal(t, checkpointCannotProceed, resolved)
+	require.Equal(t, uint64(25), barrier.Barrier.GetGlobalBarrierTs())
+	// some table not exist yet with redo is disabled.
+	barrier = schedulepb.NewBarrierWithMinTs(30)
+	redoMetaManager.enable = false
+	checkpoint, resolved = r.AdvanceCheckpoint(currentTables, time.Now(), barrier, redoMetaManager)
+	require.Equal(t, checkpointCannotProceed, checkpoint)
+	require.Equal(t, checkpointCannotProceed, resolved)
+	require.Equal(t, uint64(30), barrier.Barrier.GetGlobalBarrierTs())
 
 	span3 := spanz.TableIDToComparableSpan(3)
 	rs, err = NewReplicationSet(span3, model.Ts(5),
@@ -757,7 +768,7 @@ func TestReplicationManagerAdvanceCheckpoint(t *testing.T) {
 		}, model.ChangeFeedID{})
 	require.NoError(t, err)
 	r.spans.ReplaceOrInsert(spanRedo, rs)
-	barrier := schedulepb.NewBarrierWithMinTs(30)
+	barrier = schedulepb.NewBarrierWithMinTs(30)
 	redoMetaManager.enable = true
 	redoMetaManager.resolvedTs = 9
 	checkpoint, resolved = r.AdvanceCheckpoint(currentTables, time.Now(), barrier, redoMetaManager)

--- a/cdc/scheduler/internal/v3/replication/replication_manager_test.go
+++ b/cdc/scheduler/internal/v3/replication/replication_manager_test.go
@@ -652,22 +652,11 @@ func TestReplicationManagerAdvanceCheckpoint(t *testing.T) {
 	require.Equal(t, model.Ts(10), checkpoint)
 	require.Equal(t, model.Ts(20), resolved)
 
-	// some table not exist yet with redo is enabled.
+	// some table not exist yet.
 	currentTables.UpdateTables([]model.TableID{1, 2, 3})
-	barrier := schedulepb.NewBarrierWithMinTs(30)
-	redoMetaManager.enable = true
-	redoMetaManager.resolvedTs = 25
-	checkpoint, resolved = r.AdvanceCheckpoint(currentTables, time.Now(), barrier, redoMetaManager)
+	checkpoint, resolved = r.AdvanceCheckpoint(currentTables, time.Now(), schedulepb.NewBarrierWithMinTs(30), redoMetaManager)
 	require.Equal(t, checkpointCannotProceed, checkpoint)
 	require.Equal(t, checkpointCannotProceed, resolved)
-	require.Equal(t, uint64(25), barrier.Barrier.GetGlobalBarrierTs())
-	// some table not exist yet with redo is disabled.
-	barrier = schedulepb.NewBarrierWithMinTs(30)
-	redoMetaManager.enable = false
-	checkpoint, resolved = r.AdvanceCheckpoint(currentTables, time.Now(), barrier, redoMetaManager)
-	require.Equal(t, checkpointCannotProceed, checkpoint)
-	require.Equal(t, checkpointCannotProceed, resolved)
-	require.Equal(t, uint64(30), barrier.Barrier.GetGlobalBarrierTs())
 
 	span3 := spanz.TableIDToComparableSpan(3)
 	rs, err = NewReplicationSet(span3, model.Ts(5),
@@ -768,13 +757,58 @@ func TestReplicationManagerAdvanceCheckpoint(t *testing.T) {
 		}, model.ChangeFeedID{})
 	require.NoError(t, err)
 	r.spans.ReplaceOrInsert(spanRedo, rs)
-	barrier = schedulepb.NewBarrierWithMinTs(30)
+	barrier := schedulepb.NewBarrierWithMinTs(30)
 	redoMetaManager.enable = true
 	redoMetaManager.resolvedTs = 9
 	checkpoint, resolved = r.AdvanceCheckpoint(currentTables, time.Now(), barrier, redoMetaManager)
 	require.Equal(t, model.Ts(9), resolved)
 	require.Equal(t, model.Ts(9), checkpoint)
 	require.Equal(t, model.Ts(9), barrier.GetGlobalBarrierTs())
+}
+
+func TestReplicationManagerAdvanceCheckpointWithRedoEnabled(t *testing.T) {
+	t.Parallel()
+	r := NewReplicationManager(1, model.ChangeFeedID{})
+	span := spanz.TableIDToComparableSpan(1)
+	rs, err := NewReplicationSet(span, model.Ts(10),
+		map[model.CaptureID]*tablepb.TableStatus{
+			"1": {
+				Span:  spanz.TableIDToComparableSpan(1),
+				State: tablepb.TableStateReplicating,
+				Checkpoint: tablepb.Checkpoint{
+					CheckpointTs: model.Ts(10),
+					ResolvedTs:   model.Ts(20),
+				},
+			},
+		}, model.ChangeFeedID{})
+	require.NoError(t, err)
+	r.spans.ReplaceOrInsert(span, rs)
+
+	span2 := spanz.TableIDToComparableSpan(2)
+	rs, err = NewReplicationSet(span2, model.Ts(15),
+		map[model.CaptureID]*tablepb.TableStatus{
+			"2": {
+				Span:  spanz.TableIDToComparableSpan(2),
+				State: tablepb.TableStateReplicating,
+				Checkpoint: tablepb.Checkpoint{
+					CheckpointTs: model.Ts(15),
+					ResolvedTs:   model.Ts(30),
+				},
+			},
+		}, model.ChangeFeedID{})
+	require.NoError(t, err)
+	r.spans.ReplaceOrInsert(span2, rs)
+
+	redoMetaManager := &mockRedoMetaManager{enable: true, resolvedTs: 25}
+
+	// some table not exist yet with redo is enabled.
+	currentTables := &TableRanges{}
+	currentTables.UpdateTables([]model.TableID{1, 2, 3})
+	barrier := schedulepb.NewBarrierWithMinTs(30)
+	checkpoint, resolved := r.AdvanceCheckpoint(currentTables, time.Now(), barrier, redoMetaManager)
+	require.Equal(t, checkpointCannotProceed, checkpoint)
+	require.Equal(t, checkpointCannotProceed, resolved)
+	require.Equal(t, uint64(25), barrier.Barrier.GetGlobalBarrierTs())
 }
 
 func TestReplicationManagerHandleCaptureChanges(t *testing.T) {

